### PR TITLE
[CRIMAPP-700] Add property ownership validation

### PR DESCRIPTION
**Pull request by hiboabd**
## Description of change
Adds property ownership validation to the property and property owners screens
Property ownership needs to add up to 100 for all property owners, else a error message is shown 
Percentages entered must also be between 1 and 100 

The following scenarios are covered: 
- Displaying an error message when a percentage entered is under 1 or over 100
- Validating the percentage ownership total equals 100 if the client is the sole owner 
- Validating the percentage ownership total equals 100 if there is a partner who's percentage ownership information has been entered and no other owners 
- If a user has selected that there are other owners but has not yet entered the other owner information (as they have not yet reached the screen), the % ownership check is not undertaken 
- If a user has navigated back to the property page with incomplete property owner information, the % ownership check is not undertaken 

## Link to relevant ticket
[CRIMAPP-700](https://dsdmoj.atlassian.net/browse/CRIMAPP-700)

### After changes:

Screenshot 2024-04-22 at 10 58 09
Screenshot 2024-04-22 at 10 58 21
Screenshot 2024-04-22 at 11 04 12
Screenshot 2024-04-22 at 11 10 25

https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/51047911/3cc45fea-af31-4cac-9ff9-fdd43af7d40d

## How to manually test the feature

Scenario 1: Client is sole owner

1. Start an means tested application with an either way case type
2. Select no to partner
3. Enter a nino
4. Select no to passporting benefit
5. Go through the income, outgoings and capital screens until you get to the property type page
6. Select any property type 
7. Enter a client percentage ownership over 100 and select no to other owners
8. Enter any other data for the other questions
9. Select save and continue 
10. Confirm there is an error message displayed: `The percentage of the land your client owns must be a number between 1 and 100`
11. Change the percentage ownership figure to less than 1 
12. Confirm there is an error message displayed:  `The percentage of the land your client owns must be a number between 1 and 100` 
13. Enter a percentage over 1 and less than 100 
14. Confirm there is an error message displayed: `Percentages entered need to total 100%`
15. Enter a percentage of 100 
16. Confirm you can progress to the next screen 

Scenario 2: Client and partner owns but no other owners

1. Staying on the current application, route to the partner screen and select yes to having a partner 
2. Manually route back to the property screen and either start a new property or test with the previously started property
3. You should now see a partner percentage ownership field 
4. Enter a **partner** percentage ownership over 100 and select no to other owners
5. Enter any other data for the other questions (including a valid percentage for the client percentage ownership)
6. Select save and continue 
7. Confirm there is an error message displayed: `The percentage of the land your client's partner owns must be a number between 1 and 100`
8. Change the **partner** percentage ownership figure to less than 1 
9. Confirm there is an error message displayed:  `The percentage of the land your client's partner owns must be a number between 1 and 100` 
10. Enter a percentage that does not total for both the client and partner (e.g. 60 and 20)
11. Confirm there is an error message displayed: `Percentages entered need to total 100%`
12. Enter a total percentage ownership of 100 
13. Confirm you can progress to the next screen 

Scenario 3: Client and partner own and there are other owners

1. Manually route back to the property screen and select yes for other owners 
2. Change the total percentage ownership of the client and partner to be less than 100 
3. Confirm that you can progress to the next screen when yes is selected for other owners and the percentage ownership validation does not occur on that screen
4. Fill in the information until you reach the other property owners screen 
5. Enter an additional owner with a percentage ownership over 100
6. Select save and continue 
7. Confirm there is an error message displayed: `The percentage of the property they own must be a number between 1 and 100`
8. Change the other owner percentage ownership figure to less than 1 
9. Confirm there is an error message displayed: `The percentage of the property they own must be a number between 1 and 100`
10. Enter a percentage that does not total 100 for the client and partner and other owner(s) (e.g. 60 and 20)
11. Confirm there is an error message displayed: `Percentages entered need to total 100%`
12. Enter a total percentage ownership of 100 for all owners
13. Confirm you can progress to the next screen 

Scenario 4: Client and partner own and there are other owners with incomplete details 

1. Manually route back to the other owners screen and remove some details 
2. Select save and come back later
3. Manually route to the client's assets page `steps/capital/clients_assets`
4. You should have an incomplete property
5. Click the change link and confirm that you can select save and continue and progress back to the other property owners page from the property page 
6. On the other property owners screen, click save and continue 
7. Confirm there is an error message displayed: `Percentages entered need to total 100%` and the validation occurs on this page 

[CRIMAPP-700]: https://dsdmoj.atlassian.net/browse/CRIMAPP-700?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/forms/steps/capital/property_form.rb [hunk 1]
```diff
@@ -19,6 +19,13 @@ class PropertyForm < Steps::BaseFormObject
       validates :has_other_owners, inclusion: { in: YesNoAnswer.values }
       validates :percentage_partner_owned, presence: true, if: :include_partner?
 
+      validates_numericality_of :percentage_applicant_owned, greater_than_or_equal_to: 1.0,
+                                less_than_or_equal_to: 100.0
+      validates_numericality_of :percentage_partner_owned, greater_than_or_equal_to: 1.0,
+                                less_than_or_equal_to: 100.0, if: :include_partner?
```
Comment (hiboabd) [inline app/forms/steps/capital/property_form.rb:25]: I'm wondering if this is too restrictive.. I'm assuming the client would always have to own or part own the property but the partner could be 0?

File: app/validators/capital_assessment/property_owners_validator.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+      percentage_ownerships.sum == 100
+    end
+
+    def all_percentage_ownerships(record)
+      percentage_ownerships = record.property_owners.filter_map do |po|
+        po.percentage_owned unless po.percentage_owned.nil?
+      end
+
+      percentage_ownerships << record.record.percentage_applicant_owned
```
Comment (hiboabd) [inline app/validators/capital_assessment/property_owners_validator.rb:67]: I've not added a nil check for percentage applicant owned as it shouldn't be nil but not sure if it's worth adding as a safety precaution

File: app/validators/capital_assessment/property_ownership_validator.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      return true unless complete_property_owners?(form)
+
+      percentage_ownerships = all_percentage_ownerships(form)
+
+      percentage_ownerships.sum == 100
+    end
+
+    def all_percentage_ownerships(form)
+      percentage_ownerships = property_owners?(form) ? include_property_owners(form) : []
+      percentage_ownerships << form.percentage_applicant_owned unless form.percentage_applicant_owned.nil?
```
Comment (hiboabd) [inline app/validators/capital_assessment/property_ownership_validator.rb:25]: Here I am checking if the percentage_applicant_owned is nil but technically the validation on the field to check it's presence would catch it so it shouldn't be nil at this stage

File: app/validators/capital_assessment/property_ownership_validator.rb [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    end
+
+    def property_owners?(form)
+      property_owners = form.record.property_owners
+      form.has_other_owners&.yes? && property_owners.present? && property_owners.all?(&:complete?)
+    end
+
+    def include_property_owners(form)
+      form.record.property_owners.filter_map do |po|
+        po.percentage_owned unless po.percentage_owned.nil?
```
Comment (hiboabd) [inline app/validators/capital_assessment/property_ownership_validator.rb:42]: As I am checking for incomplete property owners then I also don't think I need the nil check here either

File: app/validators/capital_assessment/property_ownership_validator.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+      form.errors.add :percentage_applicant_owned, :invalid unless valid_ownership_total?(form)
+      form.errors.add :percentage_partner_owned, :invalid if client_has_partner(form) && !valid_ownership_total?(form)
+    end
+
+    private
+
+    def valid_ownership_total?(form)
+      return true if yet_to_add_other_owners?(form)
+      return true unless complete_property_owners?(form)
```
Comment (hiboabd) [inline app/validators/capital_assessment/property_ownership_validator.rb:16]: We want to be able to go to the next screen if property owners are incomplete (via the save and come back later button) so the user can go back to the other property owners screen and enter the ownership information

File: config/locales/en/errors.yml [hunk 6]
```diff
@@ -613,11 +613,17 @@ en:
               blank: Enter how much is left to pay on the mortgage
               not_a_number: How much there is left to pay on the mortgage must be a number, like 50
             percentage_applicant_owned:
+              invalid: Percentages entered need to total 100%
               blank: Enter the percentage of the property your client owns
-              not_a_number: The percentage of the property your client owns must be a number between 0 and 100
+              not_a_number: &property_client_percentage_error The percentage of the property your client owns must be a number between 1 and 100
```
Comment (arthurashman) [inline config/locales/en/errors.yml:618]: Error message sheet is still 0 - 100. I think user should be able to add as small as 0.01%

File: app/forms/steps/capital/property_form.rb [hunk 7]
```diff
@@ -19,6 +19,13 @@ class PropertyForm < Steps::BaseFormObject
       validates :has_other_owners, inclusion: { in: YesNoAnswer.values }
       validates :percentage_partner_owned, presence: true, if: :include_partner?
 
+      validates_numericality_of :percentage_applicant_owned, greater_than_or_equal_to: 0.0,
```
Comment (arthurashman) [inline app/forms/steps/capital/property_form.rb:22]: Wondering if this should be just greater than 0 - else they don't really own it at all, right?

Comment (arthurashman) [reply app/forms/steps/capital/property_form.rb:22]: Ignore, the partner could own 100%

File: app/forms/steps/capital/property_owner_fieldset_form.rb [hunk 8]
```diff
@@ -14,6 +14,8 @@ class PropertyOwnerFieldsetForm < Steps::BaseFormObject
         in: RelationshipType.values.map(&:to_s).push(PropertyOwner::OTHER_RELATIONSHIP)
       }
 
+      validates_numericality_of :percentage_owned, greater_than_or_equal_to: 0.0, less_than_or_equal_to: 100.0
```
Comment (arthurashman) [inline app/forms/steps/capital/property_owner_fieldset_form.rb:17]: Wondering if this should be just less than 100 - else the applicant wouldn't own it at all, right?

File: app/validators/capital_assessment/property_owners_validator.rb (hunk 2 above)
Comment (arthurashman) [reply app/validators/capital_assessment/property_owners_validator.rb:67]: Form already validates, so I think that's fine

File: app/validators/capital_assessment/property_ownership_validator.rb (hunk 3 above)
Comment (arthurashman) [reply app/validators/capital_assessment/property_ownership_validator.rb:25]: No harm

Comment (arthurashman) [reply app/validators/capital_assessment/property_ownership_validator.rb:25]: You could always run `.compact` on the array and that would remove any nil values that were added by percentage_applicant_owned.nil? or po.percentage_owned.nil?

Either way, not super necessary, but also no harm done your way

**Review by arthurashman — APPROVED** [review]
LGTM, thanks Hibo!

**Pull request merged**